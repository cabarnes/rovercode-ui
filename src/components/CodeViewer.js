import React, { Component } from 'react';
import {
  Button,
  Dialog,
  DialogContent,
  DialogTitle,
  Link,
} from '@material-ui/core';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import AceEditor from 'react-ace';
import 'brace/mode/javascript';
import 'brace/theme/monokai';

const mapStateToProps = ({ code }) => ({ code });

class CodeViewer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      open: false,
    };
  }

  stripCode = () => {
    const { code } = this.props;

    if (code.jsCode) {
      return code.jsCode.replace(/highlightBlock\(.*\);/g, '');
    }

    return '';
  }

  handleOpen = () => {
    this.setState({
      open: true,
    });
  }

  handleClose = () => {
    this.setState({
      open: false,
    });
  }

  render() {
    const { children } = this.props;
    const { open } = this.state;

    return (
      <>
        <Button variant="contained" color="primary" onClick={this.handleOpen}>
          { children }
        </Button>
        <Dialog maxWidth="md" open={open} onClose={this.handleClose}>
          <DialogTitle>
            <FormattedMessage
              id="app.code_viewer.description"
              description="Describes the programming language in the code viewer"
              defaultMessage="Here is the code in"
            />
            {' '}
            <Link href="https://en.wikipedia.org/wiki/JavaScript">
              JavaScript
            </Link>
            :
          </DialogTitle>
          <DialogContent dividers>
            <AceEditor
              mode="javascript"
              theme="github"
              readOnly
              fontSize={14}
              width="750px"
              value={this.stripCode()}
              editorProps={{ $blockScrolling: true }}
            />
          </DialogContent>
        </Dialog>
      </>
    );
  }
}

CodeViewer.propTypes = {
  code: PropTypes.shape({
    jsCode: PropTypes.string,
  }).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.instanceOf(FormattedMessage),
    PropTypes.string,
  ]).isRequired,
};

export default hot(module)(connect(mapStateToProps)(CodeViewer));
