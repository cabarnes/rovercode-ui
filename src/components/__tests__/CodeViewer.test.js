import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Button, Dialog } from '@material-ui/core';

import CodeViewer from '../CodeViewer';

describe('The CodeViewer component', () => {
  const mockStore = configureStore();

  test('renders on the page with no errors', () => {
    const store = mockStore({
      code: {
        jsCode: 'test code',
      },
    });
    const wrapper = shallow(
      <CodeViewer store={store}>
        Show Me The Code!
      </CodeViewer>,
    ).dive().dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('handles blank code with no errors', () => {
    const store = mockStore({
      code: {
        jsCode: null,
      },
    });
    const wrapper = shallow(
      <CodeViewer store={store}>
        Show Me The Code!
      </CodeViewer>,
    ).dive().dive();

    expect(wrapper.state('open')).toBe(false);

    wrapper.find(Button).simulate('click');

    expect(wrapper.state('open')).toBe(true);

    wrapper.find(Dialog).simulate('close');

    expect(wrapper.state('open')).toBe(false);
  });
});
