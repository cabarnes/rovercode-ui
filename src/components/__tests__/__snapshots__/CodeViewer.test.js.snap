// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CodeViewer component handles blank code with no errors 1`] = `
<Connect(CodeViewer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CodeViewer
    code={
      Object {
        "jsCode": null,
      }
    }
    dispatch={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Modal
      basic={true}
      centered={true}
      closeIcon={true}
      closeOnDimmerClick={true}
      closeOnDocumentClick={false}
      dimmer="blurring"
      eventPool="Modal"
      trigger={
        <Button
          as="button"
          primary={true}
        >
          Show Me The Code!
        </Button>
      }
    >
      <Portal
        closeOnDocumentClick={false}
        closeOnEscape={true}
        eventPool="Modal"
        mountNode={<body />}
        onClose={[Function]}
        onMount={[Function]}
        onOpen={[Function]}
        onUnmount={[Function]}
        openOnTriggerClick={true}
        trigger={
          <Button
            as="button"
            primary={true}
          >
            Show Me The Code!
          </Button>
        }
      >
        <Ref
          innerRef={[Function]}
        >
          <RefFindNode
            innerRef={[Function]}
          >
            <Button
              as="button"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              primary={true}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui primary button"
                    >
                      Show Me The Code!
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui primary button"
                      >
                        Show Me The Code!
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui primary button"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    Show Me The Code!
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
          </RefFindNode>
        </Ref>
      </Portal>
    </Modal>
  </CodeViewer>
</Connect(CodeViewer)>
`;

exports[`The CodeViewer component renders on the page with no errors 1`] = `
<Connect(CodeViewer)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CodeViewer
    code={
      Object {
        "jsCode": "test code",
      }
    }
    dispatch={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Modal
      basic={true}
      centered={true}
      closeIcon={true}
      closeOnDimmerClick={true}
      closeOnDocumentClick={false}
      dimmer="blurring"
      eventPool="Modal"
      trigger={
        <Button
          as="button"
          primary={true}
        >
          Show Me The Code!
        </Button>
      }
    >
      <Portal
        closeOnDocumentClick={false}
        closeOnEscape={true}
        eventPool="Modal"
        mountNode={<body />}
        onClose={[Function]}
        onMount={[Function]}
        onOpen={[Function]}
        onUnmount={[Function]}
        openOnTriggerClick={true}
        trigger={
          <Button
            as="button"
            primary={true}
          >
            Show Me The Code!
          </Button>
        }
      >
        <Ref
          innerRef={[Function]}
        >
          <RefFindNode
            innerRef={[Function]}
          >
            <Button
              as="button"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              primary={true}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui primary button"
                    >
                      Show Me The Code!
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui primary button"
                      >
                        Show Me The Code!
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui primary button"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    Show Me The Code!
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
          </RefFindNode>
        </Ref>
      </Portal>
    </Modal>
  </CodeViewer>
</Connect(CodeViewer)>
`;
